python/sglang/srt/server.py

```diff
@@ -197,9 +197,11 @@ async def v1_chat_completions(raw_request: Request):
                 request.messages, tokenize=False, add_generation_prompt=True
             )
             stop = request.stop
+            image_data = None
```

Comment:
What happen if `image_data` is None?

Reply:
I added a comment after verifying `GenerateReqInput` skips any image processing.  Basically nothing happens if `image_data` is `None`.

Reply:
Sorry let me rephrase my question a bit. What I meant was when I have a `request.messages` in `dict` type, there are two cases:

1. **chat_template_name is NOT specified**. In this case we will use `apply_chat_template`, and `image_data=None`.
2. **chat_template_name is specified**. In this case we generate conv using chat template, and `image_data=conv.image_data`.

My question is, are these two cases semantically equivalent? Or let's say, if I use the same model with temperature=0, can I get the same responses in both cases?

Reply:
When `apply_chat_template` works, they'll be the same, but really they're different.  In my testing, you can't use image input with `apply_chat_template` since it expects the `content` dict entry to always be a string and not the structured OpenAI content types I've added.  When passing in an `image_url` via the openai path, things just go wrong and the prompt comes out odd (also it won't have the `<image>` token).  

I'm not exactly sure how to best handle this, I considered modifying `request.messages` before passing it to `apply_chat_template` to be roughly equivalent to what happens in `generate_chat_conv`.  If we do that, then we could actually always extract `image_data` and then they'd be equivalent.

Reply:
Got it. Thanks for the clarification. In this case we should enforce chat template when a request has image data. For example, we return error 503 saying the server is using the chat template from the tokenizer and thus does not support image data; otherwise users would just see a weird response due to odd prompt.

Reply:
Related, I wanted to make sure I didn't break the flow where `chat_template == None`.  So I made some tweaks to the openai test suite to make sure the image input doesn't run unless requested.

I also ran everything against `meta-llama/Llama-2-7b-chat-hf` in the server since the HF chat template works.  I tested LLaVa 1.5 however the default HF chat template generates no output, which is what I'd expect given how different it is from Vicuna's template.

Reply:
Good idea on the 503.  Done, I now check for the failing request format and return a 503.